### ISSUE TITLE
source code for runs folder created after training

### ISSUE THREAD

**Issue by SONYYA7** [ultralytics/ultralytics#13216]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

i want the training logs mainly the result.png, the confusion matrix even before completing the whole training?
but those are generated after the training, how can create my own confusion matrix for the epochs it completed.

**Comment by glenn-jocher** [comment]
@SONYYA7 hello! 👋

To generate intermediate training results such as `result.png` and the confusion matrix before training completes, you can utilize the `--save-period` argument in your training command. This argument allows you to specify the frequency (in epochs) at which you'd like to save these results. For example, setting `--save-period=1` will save the results after every epoch.

Here's how you can modify your training command:

```bash
yolo detect train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 --save-period=1
```

This will save the training logs, `result.png`, and confusion matrices in the `runs/train/exp*/` directory after each epoch.

I hope this helps! Let us know if you have any more questions. 😊

**Comment by SONYYA7** [comment]
![image](https://github.com/ultralytics/ultralytics/assets/164155169/b017a746-a50d-44ab-b30d-2b57a4837672)
the sole purpose is to do inference on different dataset and get these output logs such as confusion_matrix.png, validation result logs we get . 
how to use other dataset , i.e not test or valid dataset. completely other dataset and get confusion matrix for them and these validation output logs as given in image.

please provide response soon

**Comment by SONYYA7** [comment]
yolo detect train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 --save-period=1
i did some modification, 
i can train only by this code 
from ultralytics import YOLO
config_path = '/home/ec2-user/anaconda3/envs/python3/lib/python3.10/ultralytics/cfg/models/v8/yolov8x_DW_swin_FOCUS-3.yaml'
data_path = '/home/ec2-user/SageMaker/AGAR_DATA/data.yaml'
epochs = 10
img_size = 768
workers = 4
mosaic = 1
mixup = 0.5
flipud = 0.5
fliplr = 0.5
cache = True
save = True

# Instantiate YOLO model
model = YOLO('/home/ec2-user/SageMaker/runs/detect/train56/weights/best.pt')

# Train the model
results = model.train(resume=True,patience=20,data=data_path,save_period = 5, seed=0, batch=4, epochs=epochs, imgsz=img_size, workers=workers, mosaic=mosaic, mixup=mixup, flipud=flipud, fliplr=fliplr, cache=cache, save=save )

in this how can do what u said i.e this command
yolo detect train data=coco128.yaml model=yolov8n.pt epochs=100 imgsz=640 --save-period=1

**Comment by glenn-jocher** [comment]
Hello @SONYYA7,

Great to see your progress and the modifications you've made to the training script! 👍

To integrate the `--save-period` functionality directly into your Python script, you can use the `save_period` parameter directly in the `train()` method, just as you've done in your script. It looks like you've already set `save_period=5`, which should save the intermediate results every 5 epochs.

If you want to use a completely different dataset and generate outputs like the confusion matrix for that dataset, you will need to specify the path to your new dataset in the `data` parameter of the `train()` method, just as you have done with `data_path`.

Here's a quick recap of how to adjust your script to use a different dataset for validation and generate the desired outputs:

1. Ensure your new dataset is properly formatted according to YOLOv8 requirements and referenced in a `data.yaml` file.
2. Pass this new `data.yaml` path to the `data` parameter in the `train()` method.
3. Set `save_period` as needed to generate outputs at desired intervals.

Your current setup looks correct for these adjustments. If you have any specific issues or further questions, feel free to ask!

**Comment by SONYYA7** [comment]
but this doesn't create confusion matrix for each save_period right,
i want them

**Comment by glenn-jocher** [comment]
Hello @SONYYA7,

You're correct; the `save_period` parameter will save the model weights and some other metrics, but it does not automatically generate a confusion matrix at each period.

To generate a confusion matrix for each epoch or specific intervals, you would need to modify the training loop or use a callback that evaluates the validation set and computes the confusion matrix at your desired intervals. This isn't supported directly through the standard training configuration but can be implemented with some custom coding.

If you're comfortable modifying the source code, you could add a function to compute the confusion matrix inside the training loop or as a callback function that gets called at each save period.

Let me know if you need further guidance on this or any other questions! 😊
